File: go/layer2/layer2.go
// Package layer2 issues and verifies short-lived JWT tokens used by
// hocuspocus-style long-lived connections (see design doc §10).
//
// The typical flow is: a service already authenticated by octo-server
// issues a layer-2 token good for a few minutes; the target service
// verifies the token locally on every message without another round-trip
// to octo-server. Callers still enforce the permission_epoch check
// themselves so revocations propagate within the token's remaining TTL.
//
// v1 supports HS256 only. EdDSA support is a documented v1.1 gap.
package layer2

import (
	"errors"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// DefaultTTL is the fallback TTL used when [IssueOptions.TTL] is zero. Five
// minutes matches docs-backend's COLLAB_TOKEN_TTL_SECONDS default.
const DefaultTTL = 5 * time.Minute

// MinHS256KeySize is the minimum acceptable HS256 secret length in bytes.
// golang-jwt technically accepts any []byte but shorter secrets defeat the
// whole point of HMAC — 32 bytes gives 256-bit security.
const MinHS256KeySize = 32

// Supported signature algorithms.
const (
	AlgHS256 = "HS256"
	// AlgEdDSA is accepted as a valid string but returns
	// [ErrUnsupportedAlgorithm] at Issue / Verify time — v1 is HS256-only.
	AlgEdDSA = "EdDSA"
)

// Sentinel errors returned by [IssueShortLivedToken] and
// [VerifyShortLivedToken]. Callers can use [errors.Is] for structured
// handling.
var (
	ErrUnsupportedAlgorithm = errors.New("layer2: unsupported signature algorithm (v1 supports HS256 only)")
	ErrMissingUID           = errors.New("layer2: claims missing required uid")
	ErrMissingSigKey        = errors.New("layer2: SigKey is required")
	ErrTokenExpired         = errors.New("layer2: token expired")
	ErrTokenInvalid         = errors.New("layer2: token invalid")
	ErrKeyTooShort          = fmt.Errorf("layer2: HS256 secret must be at least %d bytes", MinHS256KeySize)
)

// ShortLivedClaims is the full payload placed into a layer-2 JWT. All five
// fields are round-tripped by [IssueShortLivedToken] / [VerifyShortLivedToken].
//
// ExpiresAt is derived from IssueOptions.TTL at issue time; callers do not
// need to populate it. It IS populated on the struct returned by Verify
// so callers can render it back to the client if needed.
type ShortLivedClaims struct {
	UID             string
	DocumentName    string
	Role            string
	PermissionEpoch int64
	ExpiresAt       time.Time
}

// IssueOptions configures a call to [IssueShortLivedToken].
type IssueOptions struct {
	Claims ShortLivedClaims
	TTL    time.Duration // default DefaultTTL when zero
	SigKey []byte
	SigAlg string
}

// VerifyOptions configures a call to [VerifyShortLivedToken].
type VerifyOptions struct {
	SigKey []byte
	SigAlg string
}

// IssueShortLivedToken signs a JWT containing opts.Claims and returns the
// compact serialization. The exp claim is set to now + opts.TTL.
//
// Errors:
//   - [ErrMissingUID] when Claims.UID is empty
//   - [ErrMissingSigKey] when SigKey is nil / empty
//   - [ErrKeyTooShort] when HS256 SigKey is under MinHS256KeySize
//   - [ErrUnsupportedAlgorithm] when SigAlg is EdDSA (v1 gap) or unknown
func IssueShortLivedToken(opts IssueOptions) (string, error) {
	if opts.Claims.UID == "" {
		return "", ErrMissingUID
	}
	if len(opts.SigKey) == 0 {
		return "", ErrMissingSigKey
	}
	ttl := opts.TTL
	if ttl <= 0 {
		ttl = DefaultTTL
	}
	expiresAt := time.Now().Add(ttl)

	switch opts.SigAlg {
	case AlgHS256:
		if len(opts.SigKey) < MinHS256KeySize {
			return "", ErrKeyTooShort
		}
	case AlgEdDSA:
		return "", ErrUnsupportedAlgorithm
	default:
		return "", fmt.Errorf("%w: %q", ErrUnsupportedAlgorithm, opts.SigAlg)
	}

	claims := jwt.MapClaims{
		"uid":              opts.Claims.UID,
		"document_name":    opts.Claims.DocumentName,
		"role":             opts.Claims.Role,
		"permission_epoch": opts.Claims.PermissionEpoch,
		"exp":              expiresAt.Unix(),
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	signed, err := token.SignedString(opts.SigKey)
	if err != nil {
		return "", fmt.Errorf("layer2: sign: %w", err)
	}
	return signed, nil
}

// VerifyShortLivedToken parses tokenStr, verifies the signature and
// expiration, and returns the reconstituted claims.
//
// Errors:
//   - [ErrMissingSigKey] when SigKey is nil / empty
//   - [ErrUnsupportedAlgorithm] when SigAlg is EdDSA or unknown
//   - [ErrTokenExpired] when the exp claim is in the past
//   - [ErrTokenInvalid] for signature-mismatch, tampered payload,
//     malformed token, missing required claims (currently just uid),
//     or any other JWT-parse failure. The underlying jwt error is wrapped
//     via [errors.Unwrap] for debugging.
func VerifyShortLivedToken(tokenStr string, opts VerifyOptions) (*ShortLivedClaims, error) {
	if len(opts.SigKey) == 0 {
		return nil, ErrMissingSigKey
	}
	switch opts.SigAlg {
	case AlgHS256:
		// OK
	case AlgEdDSA:
		return nil, ErrUnsupportedAlgorithm
	default:
		return nil, fmt.Errorf("%w: %q", ErrUnsupportedAlgorithm, opts.SigAlg)
	}

	parsed, err := jwt.Parse(tokenStr, func(t *jwt.Token) (any, error) {
		// Reject "alg: none" and any non-HMAC signing method the token
		// might advertise — an attacker-controlled header can otherwise
		// swap the algorithm.
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", t.Header["alg"])
		}
		return opts.SigKey, nil
	}, jwt.WithValidMethods([]string{AlgHS256}))
	if err != nil {
		if errors.Is(err, jwt.ErrTokenExpired) {
			return nil, ErrTokenExpired
		}
		return nil, fmt.Errorf("%w: %w", ErrTokenInvalid, err)
	}
	if parsed == nil || !parsed.Valid {
		return nil, ErrTokenInvalid
	}
	claims, ok := parsed.Claims.(jwt.MapClaims)
	if !ok {
		return nil, ErrTokenInvalid
	}

	out, err := shortLivedClaimsFromMap(claims)
	if err != nil {
		return nil, err
	}
	return out, nil
}

// shortLivedClaimsFromMap reconstructs a ShortLivedClaims from the parsed
// jwt.MapClaims. Missing uid is treated as invalid; missing optional fields
// fall through to their zero values.
func shortLivedClaimsFromMap(m jwt.MapClaims) (*ShortLivedClaims, error) {
	uid, _ := m["uid"].(string)
	if uid == "" {
		return nil, fmt.Errorf("%w: missing uid", ErrTokenInvalid)
	}
	out := &ShortLivedClaims{
		UID:          uid,
		DocumentName: stringClaim(m, "document_name"),
		Role:         stringClaim(m, "role"),
	}
	// permission_epoch: JSON numbers round-trip as float64 through
	// encoding/json — convert back to int64 without losing precision for
	// realistic epoch counters.
	if v, ok := m["permission_epoch"]; ok {
		switch n := v.(type) {
		case float64:
			out.PermissionEpoch = int64(n)
		case int64:
			out.PermissionEpoch = n
		case int:
			out.PermissionEpoch = int64(n)
		}
	}
	if exp, ok := m["exp"]; ok {
		switch n := exp.(type) {
		case float64:
			out.ExpiresAt = time.Unix(int64(n), 0).UTC()
		case int64:
			out.ExpiresAt = time.Unix(n, 0).UTC()
		}
	}
	return out, nil
}

func stringClaim(m jwt.MapClaims, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}

File: ts/src/layer2.ts
/**
 * Layer-2 short-lived JWTs used by hocuspocus-style long-lived connections.
 *
 * See design doc §10 and Go SDK's layer2/layer2.go. v1 supports HS256 only;
 * EdDSA is a documented v1.1 gap and returns unsupported-algorithm at both
 * issue and verify time.
 *
 * TS uses `jose` (per R-TS-5) rather than `jsonwebtoken`. jose's
 * jwtVerify() is passed `algorithms: ['HS256']` explicitly to defend against
 * alg-confusion attacks; jose automatically rejects `alg: none`.
 */

import { SignJWT, jwtVerify, errors as joseErrors } from 'jose'
import { OctoAuthError } from './errors.js'

/** DefaultTTLSeconds matches docs-backend's COLLAB_TOKEN_TTL_SECONDS default. */
export const DEFAULT_TTL_SECONDS = 300
/** MinHS256KeySize is the minimum acceptable HS256 secret length in bytes. */
export const MIN_HS256_KEY_SIZE = 32

/** Supported signature algorithms. */
export type SigAlg = 'HS256' | 'EdDSA'
export const AlgHS256: SigAlg = 'HS256'
/** AlgEdDSA is accepted as a valid string but returns unsupported-algorithm at issue/verify time. */
export const AlgEdDSA: SigAlg = 'EdDSA'

/**
 * ShortLivedClaims is the payload placed into a layer-2 JWT. `exp` is Unix
 * seconds and is derived from IssueOptions.ttlSeconds at issue time.
 */
export interface ShortLivedClaims {
  uid: string
  documentName: string
  role: string
  permissionEpoch: number
  /** Unix seconds. Populated by verifyShortLivedToken; ignored by issue. */
  exp: number
}

/** IssueOptions configures a call to issueShortLivedToken. */
export interface IssueOptions {
  claims: Omit<ShortLivedClaims, 'exp'>
  /** TTL in seconds. Defaults to 300 (5 minutes). */
  ttlSeconds?: number
  /** Signing key bytes. Must be >= 32 bytes for HS256. */
  sigKey: Uint8Array
  sigAlg: SigAlg
}

/** VerifyOptions configures a call to verifyShortLivedToken. */
export interface VerifyOptions {
  sigKey: Uint8Array
  sigAlg: SigAlg
}

/**
 * issueShortLivedToken signs a JWT containing opts.claims and returns the
 * compact serialization. The exp claim is set to now + opts.ttlSeconds.
 *
 * Throws OctoAuthError with kinds:
 *   - invalid-credential: missing uid, missing/short sigKey
 *   - infra-failure:      signing error
 * or for EdDSA / unknown algs, unsupported-algorithm (OctoAuthError with a
 * kind not in the standard 5 — see below).
 *
 * Since our ErrorKind union doesn't include an "unsupported-algorithm"
 * literal, layer2 uses `infra-failure` for these and surfaces the reason via
 * the error message. Callers can `instanceof OctoAuthError` and inspect
 * `.message` for granular handling — matches Go's ErrUnsupportedAlgorithm
 * approach (a distinct sentinel, not an ErrorKind).
 */
export async function issueShortLivedToken(opts: IssueOptions): Promise<string> {
  if (!opts.claims || !opts.claims.uid) {
    throw new OctoAuthError('invalid-credential', 'layer2: claims missing required uid')
  }
  if (!opts.sigKey || opts.sigKey.length === 0) {
    throw new OctoAuthError('invalid-credential', 'layer2: sigKey is required')
  }
  const ttl = opts.ttlSeconds && opts.ttlSeconds > 0
    ? opts.ttlSeconds
    : DEFAULT_TTL_SECONDS
  if (opts.sigAlg === AlgHS256) {
    if (opts.sigKey.length < MIN_HS256_KEY_SIZE) {
      throw new OctoAuthError(
        'invalid-credential',
        `layer2: HS256 secret must be at least ${MIN_HS256_KEY_SIZE} bytes`,
      )
    }
  } else if (opts.sigAlg === AlgEdDSA) {
    throw new OctoAuthError(
      'infra-failure',
      'layer2: unsupported signature algorithm (v1 supports HS256 only): "EdDSA"',
    )
  } else {
    throw new OctoAuthError(
      'infra-failure',
      `layer2: unsupported signature algorithm (v1 supports HS256 only): "${String(opts.sigAlg)}"`,
    )
  }
  const nowSec = Math.floor(Date.now() / 1000)
  const expSec = nowSec + ttl
  try {
    const jwt = await new SignJWT({
      uid: opts.claims.uid,
      document_name: opts.claims.documentName,
      role: opts.claims.role,
      permission_epoch: opts.claims.permissionEpoch,
    })
      .setProtectedHeader({ alg: AlgHS256 })
      .setIssuedAt(nowSec)
      .setExpirationTime(expSec)
      .sign(opts.sigKey)
    return jwt
  } catch (err) {
    throw new OctoAuthError('infra-failure', 'layer2: sign failed', {
      cause: err,
    })
  }
}

/**
 * verifyShortLivedToken parses tokenStr, verifies the signature and
 * expiration, and returns the reconstituted claims.
 *
 * Throws OctoAuthError:
 *   - invalid-credential: missing sigKey / expired / tampered / bad shape
 *   - infra-failure:      unsupported algorithm
 */
export async function verifyShortLivedToken(
  tokenStr: string,
  opts: VerifyOptions,
): Promise<ShortLivedClaims> {
  if (!opts.sigKey || opts.sigKey.length === 0) {
    throw new OctoAuthError('invalid-credential', 'layer2: sigKey is required')
  }
  if (opts.sigAlg === AlgEdDSA) {
    throw new OctoAuthError(
      'infra-failure',
      'layer2: unsupported signature algorithm (v1 supports HS256 only): "EdDSA"',
    )
  }
  if (opts.sigAlg !== AlgHS256) {
    throw new OctoAuthError(
      'infra-failure',
      `layer2: unsupported signature algorithm (v1 supports HS256 only): "${String(opts.sigAlg)}"`,
    )
  }
  let payload: Record<string, unknown>
  try {
    // Alg-confusion defense: only HS256 is accepted; jose also rejects
    // alg: none automatically.
    const result = await jwtVerify(tokenStr, opts.sigKey, {
      algorithms: [AlgHS256],
    })
    payload = result.payload as Record<string, unknown>
  } catch (err) {
    if (err instanceof joseErrors.JWTExpired) {
      throw new OctoAuthError('invalid-credential', 'layer2: token expired', {
        cause: err,
      })
    }
    throw new OctoAuthError('invalid-credential', 'layer2: token invalid', {
      cause: err,
    })
  }
  const uid = typeof payload['uid'] === 'string' ? payload['uid'] : ''
  if (uid === '') {
    throw new OctoAuthError('invalid-credential', 'layer2: token invalid: missing uid')
  }
  const documentName =
    typeof payload['document_name'] === 'string'
      ? (payload['document_name'] as string)
      : ''
  const role = typeof payload['role'] === 'string' ? (payload['role'] as string) : ''
  const permissionEpoch = numericClaim(payload['permission_epoch'])
  const exp = numericClaim(payload['exp'])
  return {
    uid,
    documentName,
    role,
    permissionEpoch,
    exp,
  }
}

function numericClaim(v: unknown): number {
  if (typeof v === 'number') return v
  if (typeof v === 'string' && v !== '') {
    const n = Number(v)
    return Number.isFinite(n) ? n : 0
  }
  return 0
}
